[TensorExpr] Implement 'call_raw' in CUDA codegen.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57901 [TensorExpr] Implement 'call_raw' in CUDA codegen.**
* #57882 [TensorExpr] Implement 'call_raw' in IREval.
* #57881 [TensorExpr] Minor cleanup in IREval.

Differential Revision: [D28312107](https://our.internmc.facebook.com/intern/diff/D28312107)

Commit:
[TensorExpr] Implement 'call_raw' in CUDA codegen.

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28312107).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -1111,34 +1105,16 @@ void CudaCodeGen::call(const std::vector<CallArg>& args) {
-      ptr_to_args[i] = &args_data[i];
-    }
+  if (has_random_) {
+    ptr_count += 2;
```
Oh, I think I forgot to move it with the other chunk of code, it should be before the line 1109. Basically, if we have a rand, we need to add two extra arguments.

Thanks for catching!

navahgar replied:
Add a test for this case, if possible?

ZolotukhinM replied:
I'm pretty sure we already have tests for this, since this PR just moves this code around. I'm not sure why it passed, maybe because of undefined behavior (we overflowed the buffer, but it still worked)

navahgar replied:
Yeah, I was wondering the same as well.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -1111,34 +1105,16 @@ void CudaCodeGen::call(const std::vector<CallArg>& args) {
   std::vector<void*> ptr_to_args(ptr_count);
+  for (size_t i = 0; i < buffer_args.size(); i++) {
+    ptr_to_args[i] =
+        buffer_args[i].isVar() ? raw_args[i] : const_cast<void**>(&raw_args[i]);
```
This is correct and something that surprised me as well. I'll add some comments, but here is the answer I found: https://stackoverflow.com/questions/34388712/cannot-understand-how-jcuda-culaunchkernel-work

navahgar replied:
Oh wow.. that's super weird.

navahgar reviewed (approved):
LGTM

Commit:
Update on "[TensorExpr] Implement 'call_raw' in CUDA codegen."


Differential Revision: [D28312107](https://our.internmc.facebook.com/intern/diff/D28312107)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28312107).
